### ISSUE TITLE
Protect the "profilekeys" folder and files

### ISSUE THREAD

**Issue by TheMCNerd2017** [PolyMC/PolyMC#885]
### Role

Playing Minecraft

### Suggestion

The "profilekeys" folder should be protected in some way.

### Benefit

To protect users somewhat from a new false-reporting method.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

In 1.19 and newer, Mojang/Microsoft provides a private key to the player, and stores it in plaintext in the profilekeys folder. If an attacker were to get the private key file (through phishing, scams, malware, etc), they could essentially impersonate the player by signing fake chat messages with the stolen key and report them to Mojang/Microsoft, resulting in the victim getting banned.

If it's possible, PolyMC should protect the profilekeys folder and its contents in some way to protect players from this somewhat (without game modifications).

While the private keys expire after 48 hours, that is still a large enough timeframe for an attacker to generate false reports using said private key.

**Comment by txtsd** [comment]
This sounds like a Mojang/Microsoft problem. Perhaps suggest it on their bugtracker?

https://bugs.mojang.com/projects/MC/issues

**Comment by TheMCNerd2017** [comment]
> This sounds like a Mojang/Microsoft problem. Perhaps suggest it on their bugtracker?
> 
> https://bugs.mojang.com/projects/MC/issues

Yeah that would probably be the best option if it hasn't been reported on there already (it would also count as a security issue so would not even be public, even if Mojang/Microsoft would mark it as won't fix). Still though, the launcher should at least implement some extra form of protection, like deleting it on an instance exiting or excluding it entirely from instance exports/backups just in case Mojang/Microsoft doesn't care to actually fix the security flaw.

**Comment by kthchew** [comment]
Excluding it from exports by default makes sense to me, since I don't really see any reason that needs to be backed up.

Isn't the `accounts.json` that PolyMC stores a greater vulnerability than the `profilekeys`, though? `accounts.json` is slightly better in that its permissions are set so that it can only be read by the current user of the computer, but that still doesn't protect against any of the threats you mention (phishing, scams, malware, etc), and even then both `accounts.json` and `profilekeys` are generally stored in a user folder that other users don't have access to anyway. If you want to protect access to the `profilekeys` somehow then it would make sense to protect the file with the actual login tokens too.

I'm also guessing that any protections the launcher or the game can make would be mostly ineffective against malware, since the key needs to be plaintext in RAM at some point to be actually used. Maybe for phishing, but it seems a lot more likely that the user's credentials would be phished instead.

**Comment by TheMCNerd2017** [comment]
> Excluding it from exports by default makes sense to me, since I don't really see any reason that needs to be backed up.
> 
> Isn't the `accounts.json` that PolyMC stores a greater vulnerability than the `profilekeys`, though? `accounts.json` is slightly better in that its permissions are set so that it can only be read by the current user of the computer, but that still doesn't protect against any of the threats you mention (phishing, scams, malware, etc), and even then both `accounts.json` and `profilekeys` are generally stored in a user folder that other users don't have access to anyway. If you want to protect access to the `profilekeys` somehow then it would make sense to protect the file with the actual login tokens too.
> 
> I'm also guessing that any protections the launcher or the game can make would be mostly ineffective against malware, since the key needs to be plaintext in RAM at some point to be actually used. Maybe for phishing, but it seems a lot more likely that the user's credentials would be phished instead.

Yeah accounts.json can be a bigger vulnerability than the profilekeys folder, however more people are likely to know about accounts.json being sensitive rather than the profilekeys folder (due to it being a new addition in 1.19). Protecting accounts.json as well would make sense too anyways.

Also phishing for the user's login credentials is a lot less successful than phishing the account's chat private key due to much higher MS Account security that makes unauthorized logins much more difficult.

**Comment by DioEgizio** [comment]
Can this be closed since now profilekeys isn't stored in plaintext?

**Comment by TheMCNerd2017** [comment]
> Can this be closed since now profilekeys isn't stored in plaintext?

I know that's the case for newer versions (someone also stated the game doesn't cache them anymore at all), but older versions like 1.19 and late 1.19 snapshots might still store sensitive data in the profilekeys folder, unless Mojang/Microsoft went and patched those older versions as well to remove the behavior.

If you still want me to close the issue I'll do it.
